### PR TITLE
Enablement of onnxruntime for AIX and fixing issues related to big-endian platform.

### PR DESCRIPTION
**Pull request by ranjitshs**
### Description
Enablement of onnxruntime for AIX and fixing issues related to big-endian platform.

### Motivation and Context
changes in this PR contains:
1. Enablement code for building onnxruntime on AIX operating system.
2. while testing the build on AIX, we found issues related to big endian platform . More details about few of those issues can be found in  [Big endian issue: Graph Transformation Attention Fusion tests are failing #12921](https://github.com/microsoft/onnxruntime/issues/12921)

Below are list of files and the description about the change.
1.	cmake/CMakeLists.txt
[BUILDING on AIX issue] check for "IBMClang" is added for handling -Wno-unused-parameter
2.	cmake/external/onnxruntime_external_deps.cmake
[BUILDING on AIX issue]Enabling gtest_disable_pthreads for AIX
3.	cmake/onnxruntime.cmake
[BUILDING on AIX issue]
o	Blocking codes for AIX which generates generated_source.c and further requires some symbol files.
o	Putting NO AIX check for non-supported linker flags like --Xlinker
o	iconv linking
4.	cmake/onnxruntime_framework.cmake
[BUILDING on AIX issue]Putting NO AIX check for -Wl,-rpath='$ORIGIN'
5.	cmake/onnxruntime_mlas.cmake
[BUILDING on AIX issue]POWER10 releated macro/function definition .
6.	cmake/onnxruntime_providers_cpu.cmake
[BUILDING on AIX issue]Putting NO AIX check for non-supported linker flags like --Xlinker
7.	cmake/onnxruntime_unittests.cmake
[BUILDING on AIX issue]
o	Putting NO AIX check for non-supported linker flags like --Xlinker
o	Adding required libraries for AIX linker under applicatiion like onnxruntime_shared_lib_test ,onnxruntime_logging_apis etc
8.	cmake/patches/flatbuffers/flatbuffers.patch
[BUILDING on AIX issue] Handling of TypeCode in include/flatbuffers/flatbuffers.h under AIX + clang
9.	onnxruntime/contrib_ops/cpu/murmur_hash3.cc
[Big endian issue] Byte-Conversion handlling in compute() and getblock() routines
10.	onnxruntime/contrib_ops/cpu/quantization/matmul_nbits_impl.cc
[Big endian issue]  Handling of test failures . Byte swapping for quant_value.
11.	onnxruntime/core/framework/tensorprotoutils.cc
[Big endian issue]
Implementation of SetRawDataInTensorProto , ConvertRawDataInTensorProto .
 o SetRawDataInTensorProto : Wrapper for set_raw_data(). Calling ConvertRawDataInTensorProto() in big-endian system
 o ConvertRawDataInTensorProto : function used mainly on big-endian system for byte-swapping of tensor raw_data
12.	onnxruntime/core/framework/tensorprotoutils.h
[Big endian issue]
Declaration of SetRawDataInTensorProto,  ConvertRawDataInTensorProto
13.	onnxruntime/core/graph/graph.cc
[Big endian issue]
 o	Call ConvertRawDataInTensorProto for SPARSE_TENSOR type
 o	Call ConvertRawDataInTensorProto for SaveToOrtFormat
14.	onnxruntime/core/mlas/lib/platform.cpp
[BUILDING on AIX issue] POWER10 released enablement for AIX
15.	onnxruntime/core/mlas/lib/power/qgemm_kernel_power10.cpp
[BUILDING on AIX issue]Handling of __vector under AIX+clang
16.	onnxruntime/core/mlas/lib/qgemm.h
[BUILDING on AIX issue] Adding _AIX flag
17.	onnxruntime/core/mlas/lib/qlmul.cpp
[BUILDING on AIX issue] Handling of __vector under AIX+clang
18.  onnxruntime/core/optimizer/attention_fusion.cc
 [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
19.  onnxruntime/core/optimizer/compute_optimizer/shared_utils.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
20.  onnxruntime/core/optimizer/constant_folding.cc
 [Big endian issue]   Use util function SetRawDataInTensorProto, instead of set_raw_data
21.  onnxruntime/core/optimizer/embed_layer_norm_fusion.cc
 [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
22.  onnxruntime/core/optimizer/nchwc_transformer.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
23.  onnxruntime/core/optimizer/qdq_transformer/avx2_weight_s8_to_u8.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
24.  onnxruntime/core/optimizer/qdq_transformer/qdq_s8_to_u8.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
25.  onnxruntime/core/optimizer/qdq_transformer/s8_to_u8.h
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
26.  onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
27.  onnxruntime/core/optimizer/reshape_fusion.cc
 [Big endian issue]   Use util function SetRawDataInTensorProto, instead of set_raw_data
28.  onnxruntime/core/optimizer/stft_decomposition.cc
  [Big endian issue] Use util function SetRawDataInTensorProto, instead of set_raw_data
29.  onnxruntime/core/optimizer/transpose_optimization/ort_optimizer_api_impl.cc
  [Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
30.	onnxruntime/core/platform/path_lib.h
[BUILDING on AIX issue] Moving to normal function call, instead of template
31.	onnxruntime/core/platform/posix/env.cc
[BUILDING on AIX issue]Blocking syscall.h in AIX
32.	onnxruntime/core/session/inference_session.cc
 [Big endian issue]   Removing ORT_RETURN_IF_NOT, FLATBUFFERS_LITTLEENDIAN
33.	onnxruntime/test/flatbuffers/flatbuffer_utils_test.cc
  [Big endian issue] Call ConvertRawDataInTensorProto in CreateInitializer and ExternalWriteReadWithLoadInitializers
34.	onnxruntime/test/framework/sparse_kernels_test.cc
  [Big endian issue] Use util function SetRawDataInTensorProto, instead of set_raw_data
35.	onnxruntime/test/framework/tensorutils_test.cc
  [Big endian issue] Helper method ConvertEndianessForVector and call this from required place.
36.	onnxruntime/test/framework/test_tensor_loader.cc
o.  [BUILDING on AIX issue] Handling of getcwd for AIX
o.  [Big endian issue]  Bytes Swapping in run_external_data_test
37.	onnxruntime/test/onnx/main.cc
[Big endian issue] including <thread> for AIX
38.	onnxruntime/test/onnx/tensorprotoutils.cc
[Big endian issue]  Bytes swapping in UnpackTensorWithRawData
39.	onnxruntime/test/optimizer/graph_transform_test.cc
[Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
40.	onnxruntime/test/optimizer/graph_transform_test_builder.cc
[Big endian issue] Use util function SetRawDataInTensorProto, instead of set_raw_data
41.	onnxruntime/test/optimizer/graph_transform_test_builder.h
[Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
42.	onnxruntime/test/optimizer/initializer_test.cc
[Big endian issue] Use util function SetRawDataInTensorProto, instead of set_raw_data
43.	onnxruntime/test/optimizer/nchwc_optimizer_test.cc
[Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
44.	onnxruntime/test/providers/base_tester.cc
[Big endian issue]  Use util function SetRawDataInTensorProto, instead of set_raw_data
45.	onnxruntime/test/providers/cpu/generator/random_test.cc
[BUILDING on AIX issue]  Adding AIX check in MultinomialGoodCase

### CODE REVIEW

**Comment by tianleiwu** [comment]
@ranjitshs, you can fix python format by running the following at root:
```
pip install -r requirements-lintrunner.txt
pip install lintrunner
lintrunner init
lintrunner -a
```

**Comment by ranjitshs** [comment]
@tianleiwu 
Thanks for lintrunner suggestion.
Please do not initiate workflow run . I just pulled main branch. 
No change is committed yet.

**Comment by ranjitshs** [comment]
@tianleiwu 
Could you please initiate workflow execution ?

**Comment by ranjitshs** [comment]
@tianleiwu 
I have fixed conflicts from main branch. Could you please trigger build pipeline?

File: cmake/onnxruntime.cmake [hunk 1]
```diff
@@ -57,6 +57,7 @@ foreach(f ${ONNXRUNTIME_PROVIDER_NAMES})
   list(APPEND SYMBOL_FILES "${ONNXRUNTIME_ROOT}/core/providers/${f}/symbols.txt")
 endforeach()
 
+if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "AIX")
```
Comment (tianleiwu) [inline cmake/onnxruntime.cmake:60]: Any reason that generated source cannot be used in AIX?

**Comment by tianleiwu** [comment]
@ranjitshs, please follow up with the LCA comment: https://github.com/microsoft/onnxruntime/pull/21133#issuecomment-2182065619

File: cmake/onnxruntime.cmake (hunk 1 above)
Comment (ranjitshs) [reply cmake/onnxruntime.cmake:60]: > Any reason that generated source cannot be used in AIX?

Currently, I blocked the usage of generated source **generated_source.c**  and linker script **onnxruntime.lds**  file for AIX. 
AIX linker  is having different format for linker script , not the same as  linux(gnu)  linker.
Also, python script **gen_def.py** which generates these source/linkers script file need to support AIX linker script generation. 
We can consider for future enhancement .

**Comment by ranjitshs** [comment]
> @ranjitshs, please follow up with the LCA comment: [#21133 (comment)](https://github.com/microsoft/onnxruntime/pull/21133#issuecomment-2182065619)

Yes. I am working with my company legal team .

**Comment by tianleiwu** [comment]
LGTM for the code.
@ranjitshs, Any update on LCA?

**Comment by ranjitshs** [comment]
@microsoft-github-policy-service agree company="IBM"

**Comment by ranjitshs** [comment]
> @ranjitshs please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
>  > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="IBM"

**Comment by ranjitshs** [comment]
> LGTM for the code. @ranjitshs, Any update on LCA?

Done.

**Review by tianleiwu — DISMISSED** [review]

**Comment by ranjitshs** [comment]
@tianleiwu 
Thanks for previous approval.  I have fixed a recent merge conflict. 
Could you please trigger pipeline and approve it?

File: cmake/onnxruntime.cmake [hunk 2]
```diff
@@ -200,6 +210,17 @@ set(onnxruntime_INTERNAL_LIBRARIES
   onnxruntime_flatbuffers
 )
 
+if (${CMAKE_SYSTEM_NAME} MATCHES "AIX")
+  list(APPEND onnxruntime_INTERNAL_LIBRARIES  iconv)
+endif()
+
+if (onnxruntime_ENABLE_LANGUAGE_INTEROP_OPS)
```
Comment (tianleiwu) [inline cmake/onnxruntime.cmake:217]: This section of language interop is not related to AIX. Shall be removed?

Comment (ranjitshs) [reply cmake/onnxruntime.cmake:217]: sure. It got added during merge . Will remove it tomorrow.

Comment (ranjitshs) [reply cmake/onnxruntime.cmake]: Done @tianleiwu .
Please trigger pipeline once.

**Review by tianleiwu — APPROVED** [review]

**Comment by ranjitshs** [comment]
@tianleiwu 
I see one failing **orttraining-amd-gpu-ci-pipeline (Linux_Test_ubuntu)** , yesterday it was passed.  Should we re-run it ?

**Comment by tianleiwu** [comment]
> @tianleiwu I see one failing **orttraining-amd-gpu-ci-pipeline (Linux_Test_ubuntu)** , yesterday it was passed. Should we re-run it ?
Done.

**Comment by tianleiwu** [comment]
@yihonglyu, @liqunfu please take a look at mlas changes.

**Comment by ranjitshs** [comment]
> > @tianleiwu I see one failing **orttraining-amd-gpu-ci-pipeline (Linux_Test_ubuntu)** , yesterday it was passed. Should we re-run it ?
> 
> Done.

Still , it's failing. Could you please re-run.

**Comment by ranjitshs** [comment]
thanks @tianleiwu  for the approval.

[@yihonglyu](https://github.com/yihonglyu), [@liqunfu](https://github.com/liqunfu) We are targeting to merge this PR  in upcoming  v1.19, so requesting you to complete the review/approval  for mlas changes before tentatively feature complete date July 19 . Thank you in Advance.

**Review by snnn — APPROVED** [review]

**Review by liqunfu — APPROVED** [review]

**Comment by snnn** [comment]
What linker is used on AIX? It doesn't support hiding private symbols?

**Comment by ranjitshs** [comment]
> What linker is used on AIX? It doesn't support hiding private symbols?

@snnn
We are using AIX linker and AIX linker do support hiding symbols but the process of symbol management  is different than Linux. We will explore this.

**Comment by ranjitshs** [comment]
Thanks @tianleiwu  @snnn  @liqunfu @yihonglyu  for providing suggestion/help on this issue/PR.
We will monitor local CI setup and create issue if we see any build/tests failure in main branch.

**Comment by snnn** [comment]
> AIX linker do support hiding symbols but the process of symbol management is different than Linux.

Is there any public document I can take a look?

**Comment by ranjitshs** [comment]
> > AIX linker do support hiding symbols but the process of symbol management is different than Linux.
> 
> Is there any public document I can take a look?

https://www.ibm.com/docs/en/aix/7.3?topic=l-ld-command This is AIX linker man page.

For reference,  I am adding link.txt generated by cmake.  
[link.txt](https://github.com/user-attachments/files/16287901/link.txt)

In this file, we can see, cmake first creates exports.exp, then this file is passed to linker.

**Comment by snnn** [comment]
So https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/gen_def.py should be extended to support this new format.

**Comment by ranjitshs** [comment]
> So https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/gen_def.py should be extended to support this new format.

Yes . This script needs some changes for AIX support.

File: cmake/onnxruntime.cmake (hunk 1 above)
Comment (ranjitshs) [reply cmake/onnxruntime.cmake:60]: @snnn
FYI.

**Pull request merged**